src/DotVVM.Framework/Resources/Scripts/DotVVM.ts

```diff
@@ -233,9 +235,20 @@ class DotVVM {
         // handle SPA requests
         const spaPlaceHolder = this.getSpaPlaceHolder();
         if (spaPlaceHolder != null) {
-            this.domUtils.attachEvent(window, "hashchange", () => this.handleHashChange(viewModelName, spaPlaceHolder, false));
-            this.handleHashChange(viewModelName, spaPlaceHolder, true);
+            var hashChangeHandler = (initialLoad: boolean) => this.handleHashChange(viewModelName, spaPlaceHolder, initialLoad);
+
+            this.useHistoryApiSpaNavigation = <boolean>JSON.parse(<string>spaPlaceHolder.getAttribute("data-dotvvm-spacontentplaceholder-usehistoryapi"));
+            if (this.useHistoryApiSpaNavigation) {
+                hashChangeHandler = (initialLoad: boolean) => this.handleHashChangeWithHistory(viewModelName, spaPlaceHolder, initialLoad);
+            }
+
+            var spaChangedHandler = () => hashChangeHandler(false)
+            this.domUtils.attachEvent(window, "hashchange", spaChangedHandler);
+            hashChangeHandler(true);         
         }
+
+        window.addEventListener('popstate', (event) => this.handlePopState(viewModelName, event, spaPlaceHolder != null));
```

Comment:
Do we need to handle popstate when history API navigation is disabled?

Reply:
It is required when you navigate Back from normal page to SPA page, you need that navigation handle yourself.